highlighters/main/main-highlighter.zsh

```diff
@@ -1098,6 +1099,20 @@ _zsh_highlight_main_highlighter_highlight_argument()
     base_style=$REPLY
     _zsh_highlight_main_highlighter_highlight_path_separators $base_style
     highlights+=($reply)
+  elif (( $#globs )); then
+    local glob
+    local -a files; files=($~arg(NY1))
```

Comment:
`(Y1)` is new in 5.x.  In versions that don't have it, any `**` in there could be very expensive to compute (and of course it'll be recomputed every `self-insert`).

`$~foo` performs filename generation which can eval arbitrary code via `*(+foo)`.  We specifically avoid it in `_zsh_highlight_main_highlighter_expand_path`.

Reply:
Yep. The globs we match are all safe, but we don't parse any of the unsafe globs. This feature can't happen until all unsafe globs are recognized. Sorry, must have been tired.

Reply:
Not sure I follow. We shouldn't blacklist dangerous globs, but whitelist known-good ones.

Reply:
While I agree whitelisting is better, we would first have to recognize all globs in order to only allow whitelisted ones. Or we could whitelist safe globs and non-globbing characters and assume anything else is dangerous.